x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -0,0 +1,740 @@
+        final SecureString authorizationParameterValue = JwtRealm.getHeaderSchemeParameters(
+            threadContext,
+            JwtRealmSettings.HEADER_ENDUSER_AUTHORIZATION,
+            JwtRealmSettings.HEADER_ENDUSER_AUTHORIZATION_SCHEME,
+            false
+        );
+        if (authorizationParameterValue == null) {
+            return null; // Could not find non-empty SchemeParameters in HTTP header "Authorization: Bearer <SchemeParameters>"
+        }
+
+        // Get all other possible parameters. A different JWT realm may do the actual authentication.
+        final SecureString clientAuthorizationSharedSecretValue = JwtRealm.getHeaderSchemeParameters(
+            threadContext,
+            JwtRealmSettings.HEADER_CLIENT_AUTHORIZATION,
+            JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_SHARED_SECRET,
+            true
+        );
+
+        return new JwtAuthenticationToken(authorizationParameterValue, clientAuthorizationSharedSecretValue);
+    }
+
+    @Override
+    public void authenticate(final AuthenticationToken authenticationToken, final ActionListener<AuthenticationResult<User>> listener) {
+        this.ensureExpectedValueForInitialized(true);
+        if (authenticationToken instanceof JwtAuthenticationToken jwtAuthenticationToken) {
+            LOGGER.trace("Realm [{}] supports JwtAuthenticationToken.", super.name());
+            final JWSHeader jwsHeader = jwtAuthenticationToken.getJwsHeader();
+            final JWTClaimsSet jwtClaimsSet = jwtAuthenticationToken.getJwtClaimsSet();
+            final String clientAuthorizationSharedSecret = jwtAuthenticationToken.getClientAuthorizationSharedSecret().toString();
+
+            // Filter steps (before any validation)
+
+            // 1. Skip JWT if signature algorithm does not match any of the signature algorithms allowed by this realm.
+            final JWSAlgorithm jwsSignatureAlgorithm = jwsHeader.getAlgorithm();
+            if ((jwsSignatureAlgorithm == null) || (this.allowedSignatureAlgorithms.contains(jwsSignatureAlgorithm.getName()) == false)) {
+                final String msg = String.format(
+                    Locale.ROOT,
+                    "Realm [%s] does not allow signature algorithm [%s]. Allowed signature algorithms are %s.",
+                    super.name(),
+                    jwsSignatureAlgorithm,
+                    this.allowedSignatureAlgorithms
+                );
+                LOGGER.debug(msg);
+                listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                return;
+            }
+            LOGGER.debug(
+                "Realm [%s] allows signature algorithm [%s]. Allowed signature algorithms are %s.",
+                super.name(),
+                jwsSignatureAlgorithm,
+                this.allowedSignatureAlgorithms
+            );
+
+            // 2. Skip JWT if issuer does not match the issuer allowed by this realm.
```

Comment:
Why do we them in this order?
I think issuer then algorithm is better, but perhaps there's a good counter-argument.

If you have 2 JWT realms with different issuers and different signature algorithms, and you turn on debug logging, then it makes sense to see log messages saying that 1 of those realms skipped JWTs because they're for the other issuer. You can probably follow what's going on there, and why it would happen.
But if you see log messages about the signature algorithms not matching, then it looks like you have a configuration problem. But you don't - it's totally normal it's just a consequence of the wrong realm setting the JWT first.

So, in the absence of a counter argument, I think issuer validation should be first.

Reply:
I will re-order them as issuer, audiences, and algorithms.

Reply:
Done